cvise/passes/clanghints.py

```diff
@@ -0,0 +1,92 @@
+        if parent is None:
+            return None
+        return ClangState(clang_std, parent.hint_count, parent.hints_file_path, parent.binary_state)
+
+
+class ClangHintsPass(HintBasedPass):
+    """A pass that performs reduction using the hints produced by the clang_delta tool.
+
+    Implementation-wise, we don't use default new/advance/advance_on_success implementation from the base class, because
+    we want to brute-force Clang's `--std=` parameter that maximizes the generated set of hints. This requires having
+    special logic in new() and carrying over some extra information from new() to advance_on_success() throughout all
+    advance() calls.
+    """
+
+    def check_prerequisites(self):
+        return self.check_external_program('clang_delta')
+
+    def new(self, test_case, tmp_dir, **kwargs):
+        # Choose the best standard unless the user provided one.
+        std_choices = [self.user_clang_delta_std] if self.user_clang_delta_std else CLANG_STD_CHOICES
+        best_std = None
+        best_hints = []
+        for std in std_choices:
+            start = time.monotonic()
+            hints = self.generate_hints_for_standard(test_case, std)
+            took = time.monotonic() - start
+            # prefer newer standard if the # of instances is equal
+            if len(hints) >= len(best_hints):
+                best_std = std
+                best_hints = hints
+            logging.debug('available transformation opportunities for %s: %d, took: %.2f s' % (std, len(hints), took))
+        logging.info('using C++ standard: %s with %d transformation opportunities' % (best_std, len(best_hints)))
+
+        # Let the parent class complete the initialization, but create our own state to remember the chosen standard.
+        hint_state = self.new_from_hints(best_hints, tmp_dir)
+        return ClangState.wrap(hint_state, best_std)
+
+    def advance(self, test_case, state):
+        new_state = super().advance(test_case, state)
+        # Re-attach the remembered standard.
+        return ClangState.wrap(new_state, state.clang_std)
+
+    def advance_on_success(self, test_case, state):
+        # Keep using the same standard as the one chosen in new() - repeating the choose procedure on every successful
+        # reduction would be too costly.
+        hints = self.generate_hints_for_standard(test_case, state.clang_std)
+        new_state = self.advance_on_success_from_hints(hints, state)
+        return ClangState.wrap(new_state, state.clang_std)
+
+    def generate_hints_for_standard(self, test_case, std):
+        cmd = [
+            self.external_programs['clang_delta'],
+            f'--transformation={self.arg}',
+            f'--std={std}',
+            '--generate-hints',
+            test_case,
+        ]
+
+        logging.debug(shlex.join(str(s) for s in cmd))
+
+        hints = []
+        with subprocess.Popen(cmd, stdout=subprocess.PIPE, text=True) as proc:
```

Comment:
One missing thing, compared to `clangbinarysearch.py`, is that we don't set a timeout here. It's crucial for the cases when building Clang AST hangs.

I'm planning to address this in a follow-up - I want to implement it in a more robust way than the original code's hardcoded 10 seconds (https://github.com/marxin/cvise/blob/ed1cba72a7c3227841527e73cdf4f15487c09241/cvise/passes/clangbinarysearch.py#L13).